### DISCUSSION TITLE
Build a fluid block from a moving emitter

### DISCUSSION THREAD

**Discussion by ssungho98** [InteractiveComputerGraphics/SPlisHSPlasH#262] [Q&A] [answered]
Hi! I would like to make a line of fluid block constructed by a fluid particles generated from a mobile emitter.

By using python scripts, I was able to change the emitter position. However, I could not do it repeatedly.

My ultimate goal is to get position data from Gazebo and put the emitter at that position so that it can move in every time frame.

Does anyone have tried this before? I am open to any suggestion!

**Comment by janbender** [comment]
If you take a look at the example "callbacks.py", there is a function "time_step_callback()" which is called after each simulation step. You can try to move the emitter in such a function.

**Reply by ssungho98** [reply]
Thank you. I am still figuring out how to code this one. How can I get a time step as a variable?

**Reply by ssungho98** [reply]
Additionally, what is the correct way to pass arguments to callback function?

**Comment by ssungho98** [comment]
Hi, @janbender!  I have another question.

I edited complex_scene.py to linearly move the emitter in x-direction.

```
def time_step_callback():
	scene = sph.Exec.SceneConfiguration.getCurrent().getScene()
	xpos=scene.emitters[0].x[0]
	# update position
	scene.emitters.append(Scenes.EmitterData(id='Fluid2', x=[xpos+1, 0.0, 0.0]))
	print(scene.emitters)
```

The above callback function was not updated to the simulation.

How can I fix this? Thank you so much always for your help.

**Reply by ssungho98** [reply]
If I write the code as follows, I found that the printed position was updated, yet it was not reflected on the simulation.
```
def time_step_callback():
	scene = sph.Exec.SceneConfiguration.getCurrent().getScene()
	xpos=scene.emitters[0].x[0]
	# update position
	scene.emitters[0].x=[xpos+0.01, 0.0, 0.0]
	print(scene.emitters[0].x)
```

**Comment by janbender** [comment, marked as answer]
Hi, 

This won't work since "scene" just contains the data from the scene file. This is read once at the start of the simulator and then the simulation is initialized. Modifying this data afterwards has no effect. 

You have to modify the emitter directly. Each FluidModel (you can get it from the Simulation) has an emitter system (fluidModel.getEmitterSystem()). Each emitter system has some emitters (getEmitters()). If you have the emitter, you can call setPosition(). 

See also: 
https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/master/pySPlisHSPlasH/EmitterModule.cpp#L18

Hope that helps.

**Reply by ssungho98** [reply]
Thank you so much! Now I finally am able to update the emitter position. Still the problem is, although the particles are generated in the moving position, the emitter block in the simulation is not updated. The emitter block is stationary at the starting position. Do you know why it's not been reflected?

**Comment by janbender** [comment]
The box around the emitter is a static boundary object. If you not absolutely need it, you could comment out this line: 
https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/master/Simulator/SimulatorBase.cpp#L1337

Then it should not be added to the scene.

**Reply by ssungho98** [reply]
Thank you! I will do so.

**Reply by ssungho98** [reply]
Hi! The box does not disappear even though I commented out the line `// scene.boundaryModels.push_back(emitterBoundary);`

**Reply by janbender** [reply]
This is strange. I just tested it with the Emitter.json scene and for me it works.
